### PR TITLE
feat: ✨ add inline validation + improve initial validation

### LINKED ISSUE #631: fix: 🤔 optimize initial validation in form components
>NOTE: 
This applies to all form elements / input components. The description below only handles the `sd-input` component as an example.
The fix should be implemented into all necessary components (`sd-switch`, `sd-checkbox`, `sd-radio-group`, `sd-input`, `sd-textarea` & `sd-select`).

## Current behavior
1. When using a required `sd-input` without any user interaction, the field is marked as red. This isn't very friendly UX: https://codepen.io/mariohamann/pen/KKJxXQx
2. When using a required `sd-input` with a prefilled value, the input still shows an error state (red). Only when interacting with the input, the components switches to the success state (green): https://codepen.io/mariohamann/pen/poGOWdO

## Expected behavior
There should be no validation indication until the user interacted with the field. It should never show an invalid/valid state as long as the user didn't interact.

## Technical Information
See https://shoelace.style/getting-started/form-controls#styling-invalid-form-controls for more details about the expected behavior. Especially about the attributes "data-valid" and "data-user-valid" to be used for visual validation states.

## Use-case:
The user builds a contact form, where the (logged in) end-user data is prefilled. The end-user then can edit the values or leave them like they are.
## DoR
- [x] Item has business value
- [x] Item has been estimated by the team
- [x] Item is clear and well-defined
- [x] Item dependencies have been identified

## DoD
- [ ] Documentation has been created/updated (if applicable)
- [ ] Migration Guide has been created/updated (if applicable)
- [ ] Relevant E2E tests (Features, A11y, Bug fixes) are created/updated
- [ ] Relevant stories (Features, A11y) are created/updated
- [ ] Implementation works successfully on `feature` branch

-------------------------------------------------------------------------------------------------------------------------
As of 05.12.2023 we are combining the following ticket with this one: 
https://github.com/solid-design-system/solid/issues/501

We will use the `data-user-valid` and `data-user-invalid` attributes provided by the `formController` to condition our component styles based on form validity state and user interaction.

## Requirements: for each listed component...

- [ ] add private boolean state `showInvalidStyle`
- [ ] add private boolean state `showValidStyle` (if component has valid style)
- [ ] condition validity styles on `this.showValidStyle` and `this.showInvalidStyle`
- [ ] render div at bottom with id "error-message" for displaying inline error
- [ ] add query errorMessage for consistent access
- [ ] update handleInvalid method to display browser validation messages in errorMessage.textContent property

## Components
- [x] `sd-switch` 
- [x] `sd-input`
- [x] `sd-textarea`
- [x] `sd-select`
- [x]  `sd-checkbox`
- [x]  `sd-radio-group` #641 

## Delimitations
- `sd-checkbox-group` is handled in #640

### LINKED ISSUE #641: fix: 🤔 align `sd-radio/-group` invalid state with WCAG 2.1
## Current behavior
The `sd-radio` is only colored in red, in case the component gets an `invalid` state. This does not align with [WCAG 1.4.1 Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html).
Description can be found [here](https://www.smashingmagazine.com/2023/02/guide-accessible-form-validation/#the-color-trap).
## Expected behavior
The component shows an additional `error-text`.
## Open Questions
- [ ] should this be added to the `sd-radio` or `sd-radio-group`? The `sd-radio` is handling the error state but an additional error-text to each radio item does not make any sense. Adding the `error-text` to the group does not fit either as the group currently does not handle any error state.
- [ ] Where do we handle this in the `sd-radio-button-group` or can we forget this component as there will not be any "error-state" as the component is always valid?

## DoR
- [ ] Item has business value
- [ ] Item has been estimated by the team
- [ ] Item is clear and well-defined
- [ ] Item dependencies have been identified

## DoD
- [ ] Documentation has been created/updated (if applicable)
- [ ] Migration Guide has been created/updated (if applicable)
- [ ] Relevant E2E tests (Features, A11y, Bug fixes) are created/updated
- [ ] Relevant stories (Features, A11y) are created/updated
- [ ] Implementation works successfully on `feature` branch

### PR DESCRIPTION
**Pull request by abudd1094**
Closes #631 
Closes #641 additionally

### CODE REVIEW

File: packages/components/src/internal/form.ts
Comment (abudd1094) [inline packages/components/src/internal/form.ts:363]: more specifically, event.preventDefault() can either happen here, or it needs to be implemented in every "handleInvalid" method to prevent the browser validation messages from showing up in their overlay style

File: packages/components/src/components/input/input.ts
Comment (mariohamann) [reply packages/components/src/components/input/input.ts:265]: Makes it much clearer for me, what actually changes. Please adapt everywhere.

File: packages/components/src/components/select/select.stories.ts
Comment (abudd1094) [reply packages/components/src/components/select/select.stories.ts:54]: What do you mean by filter?  It is an object so I opted for delete keyword. BUT, then linter complains about the "any" type (which is used everywhere else...).

**Comment by abudd1094** [comment]
> This change feels SO nice. I really enjoy it.
> 
> Two things:
> 
> 1. Is it too much work to implement it for the radio-group as well?
> 2. Could you make a sample + docs showing how to use `setCustomValidity` like shown here? https://shoelace.style/getting-started/form-controls#custom-error-messages
> 
> As we don't provide an "error" state and people use the help-text anymore, this would really help for migrating, I believe. What do you think?

Sure I can add a sample.  It's not too much work, but the strategy is not clear like @karlbaumhauer mentioned.  Group vs. individual radios.  It does get a bit more complicated and some of the same logic starts to not work, I got some strange error when I tried it, which was another reason we deferred that to its own ticket.

File: packages/components/src/components/select/select.stories.ts
Comment (abudd1094) [reply packages/components/src/components/select/select.stories.ts:54]: Ok I'll give it a shot

File: packages/components/src/components/input/input.ts
Comment (abudd1094) [reply packages/components/src/components/input/input.ts:265]: Ok I'll try to make it clearer

**Comment by mariohamann** [comment]
> Sure I can add a sample. It's not too much work, but the strategy is not clear like @karlbaumhauer mentioned in #641 . It does get a bit more complicated and some of the same logic starts to not work, I got some strange error when I tried it, which was another reason we deferred that to its own ticket.

Okay. So from my understanding it should only happen in the group and I think that's what happens here, when I select something and then remove the value via browser, too:

CleanShot 2023-12-08 at 15 50 04@2x

The ticket says that the single radios handle the validation, what is not correct and the screenshot pretty clearly shows that.

As you've changed the `form.ts` now, this would be a mandatory change from my perspective, as the browser popup now doesn't show up any more.

File: packages/components/src/components/textarea/textarea.ts
Comment (mariohamann) [reply packages/components/src/components/textarea/textarea.ts:398]: You're right. Leave it like it is. We could handle this in a different ticket if needed.

What do you think of providing a single class for the error message then which could be used for all inputs as well?

Comment (abudd1094) [reply packages/components/src/components/textarea/textarea.ts:398]: I like that idea, was not sure where to implement it.  Will inspect and implement something.

File: packages/components/src/components/select/select.stories.ts
Comment (abudd1094) [reply packages/components/src/components/select/select.stories.ts]: I've gone with casting args in each story where "open-attr" needs to be removed: `args: { 'open-attr'?: string }`

Simple and effective

**Comment by abudd1094** [comment]
@mariohamann what happened to this PR?

**Comment by mariohamann** [comment]
Sorry. My mistake. I closed the wrong ticket...

**Comment by abudd1094** [comment]
@mariohamann  ok, no problem... 

Pull request has been updated and is ready for review.  Refer to [the ticket ](https://github.com/solid-design-system/solid/issues/631) for information about the chosen strategy.

Check out my branch fix/optimize-validation, start the storybook, and look at the story in sd-select: "Sample: Solid Form".  Click around, try to break it, see how it feels.  It should be removed from sd-select before merging to main.  Is there somewhere else we could move it to keep it around?  It will be helpful for developers trying to understand / implement our form components.

## What changed?

### form.styles.ts is removed completely 
Likewise, all classes prefixed with "form-control" are removed from the individual components.  

Why?  Because they were inconsistently applying styles that either did nothing, were overwritten by tailwind, or no longer relevant.  We are taking the approach of Tailwind first, CSS if necessary.  The mere presence of these classes is confusing and makes our code unnecessarily bloated.  We keep the CSS form-control parts and remove the classes. This makes it clear, we use Tailwind in each component, period.

### form.ts takes care of validation styles
Each component has new states showInvalidStyle (and sometimes showValidStyle) which can be mutated directly from the formControlController via its host property.  This takes care of all styling logic in one central place.

### inline styles use ternary expressions
We started to have a mix of inline styles (e.g. in checkbox, switch, radio-group, radio). Some would use `(condition && condition && style) || (condition && condition && style) ...` they now use ternary expressions.  Why?  It is objectively more readable and less cluttered.

For further context see my individual comments.

File: packages/components/src/components/checkbox/checkbox.ts
Comment (abudd1094) [inline packages/components/src/components/checkbox/checkbox.ts:39]: I needed to annotate the formControlControllers to have my new methods work

File: packages/components/src/components/input/input.stories.ts
Comment (abudd1094) [inline packages/components/src/components/input/input.stories.ts:596]: This is a more efficient way to check for form validity

File: packages/components/src/components/input/input.stories.ts
Comment (abudd1094) [inline packages/components/src/components/input/input.stories.ts:738]: Consistent naming is important

File: packages/components/src/components/input/input.ts
Comment (abudd1094) [inline packages/components/src/components/input/input.ts:283]: this will load the default browser validation messages into the inline error

File: packages/components/src/components/input/input.ts
Comment (abudd1094) [inline packages/components/src/components/input/input.ts:657]: Design explicitly states to use a space character and an asterisk, we don't need margin left in any of these cases.

File: packages/components/src/components/radio-group/radio-group.ts
Comment (abudd1094) [inline packages/components/src/components/radio-group/radio-group.ts:60]: This is deprecated with the new logic in form.ts.  We should have a consistent approach to validity.  None of the other components use an 'invalid' property.

File: packages/components/src/components/radio-group/radio-group.ts
Comment (abudd1094) [inline packages/components/src/components/radio-group/radio-group.ts:67]: Design explicitly state we should have default values for labels to enforce accessibility standards, I have applied this.

File: packages/components/src/components/radio-group/radio-group.ts
Comment (abudd1094) [inline packages/components/src/components/radio-group/radio-group.ts:207]: Shoelace used handleLabelClick, but this is now used in form.ts to focus when this element is the first invalid element in a form.  focus makes more sense, makes it more consistent, and makes it useable in the form method

File: packages/components/src/internal/form.ts
Comment (abudd1094) [inline packages/components/src/internal/form.ts:212]: Check out the Solid Form sample for context.  Without this alteration, the inline error message will not show up on an invalid submission.

File: packages/components/src/internal/form.ts
Comment (abudd1094) [inline packages/components/src/internal/form.ts:329]: As far as I can tell, there is no downside to rendering the error message like this. All designs match and the attributes are always the same.  Instead of providing a class, I took this approach.  Thoughts?

**Comment by abudd1094** [comment]
> > This change feels SO nice. I really enjoy it.
> > Two things:
> > 
> > 1. Is it too much work to implement it for the radio-group as well?
> > 2. Could you make a sample + docs showing how to use `setCustomValidity` like shown here? https://shoelace.style/getting-started/form-controls#custom-error-messages
> > 
> > As we don't provide an "error" state and people use the help-text anymore, this would really help for migrating, I believe. What do you think?
> 
> Sure I can add a sample. It's not too much work, but the strategy is not clear like @karlbaumhauer mentioned in #641 . It does get a bit more complicated and some of the same logic starts to not work, I got some strange error when I tried it, which was another reason we deferred that to its own ticket.

@mariohamann 
Looking back at this comment, adding a sample for this stuff makes sense, but it does not make sense to duplicate in every form component.  Moreover a lot of those stories could be reduced into an outer level Folder named "Form Controls" or something.  Notice Shoelace does that under their "Getting Started" section.  This would be a good place for the "Solid Form" story.  I don't want to put in the work to research and implement that without prior approval.  Please give me feedback or get in touch and we can go from there.

**Review by mariohamann — CHANGES_REQUESTED** [review]

File: packages/components/src/components/radio-group/radio-group.ts
Comment (mariohamann) [reply packages/components/src/components/radio-group/radio-group.ts:54]: I would like to bring in the "invalid" wording more, that it's clearer that this is connected to form validation.

File: packages/components/src/components/input/input.ts
Comment (abudd1094) [reply packages/components/src/components/input/input.ts:639]: Daniel and Özlem have been doing that elsewhere and it has been previously reviewed and approved.  I thought I was improving consistency by adopting that.  We should do one or the other.  
image

File: packages/components/src/components/radio-group/radio-group.ts
Comment (abudd1094) [reply packages/components/src/components/radio-group/radio-group.ts:54]: Seems like 'validationMessage' is the standard:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/validationMessage

File: packages/components/src/components/radio/radio.ts
Comment (abudd1094) [reply packages/components/src/components/radio/radio.ts:152]: -_-

File: packages/components/src/components/radio-group/radio-group.ts
Comment (abudd1094) [reply packages/components/src/components/radio-group/radio-group.ts:54]: Because it is an element and not a string, and it only appears for errors, the name "inlineError" would make sense to me

Comment (abudd1094) [reply packages/components/src/components/radio-group/radio-group.ts]: OK we went with invalidMessage

**Review by mariohamann — APPROVED** [review]
Let's ship it! I adapted the PR title, hope that's fine like this. :)

**Pull request merged**
